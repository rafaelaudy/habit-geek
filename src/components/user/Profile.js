//

import React, { useState } from "react";
import { navigate } from "@reach/router";
import { useTranslation } from "react-i18next";

import "./Profile.scss";
import avatar1 from "@rafael.audy/habit-geek-utils/imgs/avatars/001-burglar.svg";
import avatar2 from "@rafael.audy/habit-geek-utils/imgs/avatars/002-woman.svg";
import avatar3 from "@rafael.audy/habit-geek-utils/imgs/avatars/003-superhero.svg";
import avatar4 from "@rafael.audy/habit-geek-utils/imgs/avatars/004-robot.svg";
import avatar5 from "@rafael.audy/habit-geek-utils/imgs/avatars/005-dragon.svg";
import avatar6 from "@rafael.audy/habit-geek-utils/imgs/avatars/006-cyborg.svg";

const avatars = [avatar1, avatar2, avatar3, avatar4, avatar5, avatar6];

const Profile = ({ name: initialName, avatar, registerUser }) => {
  const [name, setName] = useState(initialName);
  const [selectedAvatar, setAvatar] = useState(avatar ? avatar : avatar1);
  const [validityChecked, setValidityChecked] = useState(false);
  const { t } = useTranslation();

  const register = event => {
    event.preventDefault();
    setValidityChecked(true);

    if (event.target.checkValidity()) {
      registerUser(name, selectedAvatar);
      navigate("/habits");
    }
  };

  return (
    <form
      className={`mobile-size-container ${
        validityChecked ? "was-validated" : ""
      }`}
      onSubmit={register}
      noValidate
    >
      <h2>{t("profile-title")}</h2>
      <div className="mb-3">
        <label htmlFor="profile-name">{t("profile-name")}</label>
        <input
          id="profile-name"
          className="form-control"
          value={name}
          onChange={e => setName(e.target.value)}
          required
        ></input>
        <div className="invalid-feedback">{t("profile-name-validation")}</div>
      </div>
      <div className="mb-3">
        <label htmlFor="profile-avatar">{t("profile-avatar")}</label>
        <div className="input-group profile__avatar-container">
          {avatars.map((avatar, index) => (
            // eslint-disable-next-line jsx-a11y/no-noninteractive-element-interactions, jsx-a11y/click-events-have-key-events
            <img
              key={`avatar-${index}`}
              onClick={() => setAvatar(avatar)}
              src={avatar}
              alt={t("profile-avatar-alt-desc")}
              className={`profile__avatar ${
                avatar === selectedAvatar ? "profile__avatar--selected" : ""
              }`}
            ></img>
          ))}
        </div>
      </div>
      <hr className="mb-4" />
      <button type="submit" className="btn btn-primary btn-lg btn-block">
        {t("profile-save")}
      </button>
    </form>
  );
};

export default Profile;
