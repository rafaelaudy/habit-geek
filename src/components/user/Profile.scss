.profile {
  &__avatar-container {
    display: flex;
    justify-content: space-around;
    height: 50px;
  }

  &__avatar {
    width: 40px;
    height: 50px;
    border-radius: 50%;
  }

  &__avatar--selected {
    transition-property: width;
    transition-duration: 0.2s;
    border: 3px solid var(--primary);
    outline: none;
    width: 50px;
  }
}
