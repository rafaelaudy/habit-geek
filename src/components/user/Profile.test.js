import React from "react";
import { navigate } from "@reach/router";
import { shallow } from "enzyme";

import Profile from "./Profile";

let fakeEvent;

jest.mock("@reach/router", () => ({
  navigate: jest.fn()
}));

describe("Profile component", () => {
  afterAll(() => {
    jest.unmock("@reach/router");
  });

  beforeEach(() => {
    jest.clearAllMocks();
    fakeEvent = {
      preventDefault: jest.fn(),
      target: { checkValidity: jest.fn().mockReturnValue(true) }
    };
  });

  it("Renders static elements", () => {
    const profile = shallow(<Profile avatar="002-woman.svg" name="rafa" />);
    expect(profile).toMatchSnapshot();
  });

  it("Register and navigates to Dashboard", () => {
    const registerUserAction = jest.fn();
    const profile = shallow(<Profile registerUser={registerUserAction} />);
    profile
      .find("#profile-name")
      .simulate("change", { target: { value: "Rafa" } });
    profile
      .find(".profile__avatar")
      .at(0)
      .simulate("click");
    profile.find("form").simulate("submit", fakeEvent);
    expect(registerUserAction).toHaveBeenCalled();
    expect(registerUserAction).toHaveBeenCalledWith("Rafa", "001-burglar.svg");
    expect(navigate).toHaveBeenCalledWith("/habits");
  });

  it("validates forms", () => {
    const registerUserAction = jest.fn();
    fakeEvent.target.checkValidity.mockReturnValueOnce(false);
    const profile = shallow(<Profile registerUser={registerUserAction} />);
    profile.find("form").simulate("submit", fakeEvent);
    expect(profile.find(".was-validated").length).toBe(1);
    expect(fakeEvent.preventDefault).toHaveBeenCalledTimes(1);
    expect(fakeEvent.target.checkValidity).toHaveBeenCalledTimes(1);
    expect(registerUserAction).toHaveBeenCalledTimes(0);
    expect(navigate).toHaveBeenCalledTimes(0);
  });
});
