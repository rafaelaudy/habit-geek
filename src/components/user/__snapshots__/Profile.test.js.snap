// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Profile component Renders static elements 1`] = `
<form
  className="mobile-size-container "
  noValidate={true}
  onSubmit={[Function]}
>
  <h2>
    profile-title
  </h2>
  <div
    className="mb-3"
  >
    <label
      htmlFor="profile-name"
    >
      profile-name
    </label>
    <input
      className="form-control"
      id="profile-name"
      onChange={[Function]}
      required={true}
      value="rafa"
    />
    <div
      className="invalid-feedback"
    >
      profile-name-validation
    </div>
  </div>
  <div
    className="mb-3"
  >
    <label
      htmlFor="profile-avatar"
    >
      profile-avatar
    </label>
    <div
      className="input-group profile__avatar-container"
    >
      <img
        alt="profile-avatar-alt-desc"
        className="profile__avatar "
        key="avatar-0"
        onClick={[Function]}
        src="001-burglar.svg"
      />
      <img
        alt="profile-avatar-alt-desc"
        className="profile__avatar profile__avatar--selected"
        key="avatar-1"
        onClick={[Function]}
        src="002-woman.svg"
      />
      <img
        alt="profile-avatar-alt-desc"
        className="profile__avatar "
        key="avatar-2"
        onClick={[Function]}
        src="003-superhero.svg"
      />
      <img
        alt="profile-avatar-alt-desc"
        className="profile__avatar "
        key="avatar-3"
        onClick={[Function]}
        src="004-robot.svg"
      />
      <img
        alt="profile-avatar-alt-desc"
        className="profile__avatar "
        key="avatar-4"
        onClick={[Function]}
        src="005-dragon.svg"
      />
      <img
        alt="profile-avatar-alt-desc"
        className="profile__avatar "
        key="avatar-5"
        onClick={[Function]}
        src="006-cyborg.svg"
      />
    </div>
  </div>
  <hr
    className="mb-4"
  />
  <button
    className="btn btn-primary btn-lg btn-block"
    type="submit"
  >
    profile-save
  </button>
</form>
`;
